### ISSUE TITLE
BUG: optimize.least_squares giving poor result compared to optimize.leastsq and optimize.curve_fit

### ISSUE THREAD

**Issue by leeping** [scipy/scipy#19459]
### Describe your issue.

I am seeing some poor results when using `optimize.least_squares` to fit a 1-D function.  The fitting result is highly dependent on the initial guess.  I understand these issues can be easily caused by a poor initial guess, but I do not believe that is what's going on here.

In the attached screenshot, the top left panel shows the poor fitting result if the peak position of the initial model is shifted slightly to the right.  The top right panel uses different parameters where the peak position is shifted slightly to the left and that gives the expected result.  I also ran tests using `optimize.leastsq` and `optimize.curve_fit` and they do not have this issue; one obtains the expected result for both initial guesses.  I also know that `optimize.curve_fit` can give a consistently good fit for different data, as I have about 100 experimental peaks in this dataset with different positions, areas, and widths and I was able to fit them all successfully with `optimize.curve_fit` (but not `optimize.least_squares`).

Other things I tried, which did not work for me:

- Switching the optimization algorithm to `lm` or `dogbox`
- Setting `max_nfev` to a very large number (1e6)
- Trying several values for `diff_step` (1e-6 or 1e-8)

Changing these settings will sometimes give a different optimization result, but still not correct.

I have attached a screenshot and the codes to reproduce the error as a script and in Jupyter Notebook format.  The data file is also attached.

Thanks for your attention to this issue!  SciPy is extremely valuable to my research and this is the first issue I have had to report in >10 years as a user.
 
![Screenshot 2023-11-01 094635](https://github.com/scipy/scipy/assets/1441560/56d03a7d-e550-4e77-ac63-e3842ca3dd38)

[test.ipynb.zip](https://github.com/scipy/scipy/files/13229723/test.ipynb.zip)

[R.csv.zip](https://github.com/scipy/scipy/files/13229811/R.csv.zip)

### Reproducing Code Example

```python
#!/usr/bin/env python

import numpy as np
import scipy
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from scipy.optimize import least_squares, leastsq, curve_fit

# The bug is observed with SciPy version 1.11.3
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()

# Load the data and reshape into 1-D arrays.
X_array = np.loadtxt('R.csv',skiprows=1,delimiter=',')
deltas = X_array[:,0].flatten()
signal = X_array[:,1].flatten()

# Lorentzian model function. x0 = center; A = area; sigma = width
def lorentzian(x, x0, A, sigma):
    return A * sigma / ((x-x0)**2 + sigma**2)

# Function to calculate the residual.
def calc_residual(params):
    model = lorentzian(deltas, params[0], params[1], params[2])
    residual = (model - signal)
    return residual

# Two sets of initial parameters.
p_init0=np.array([-62.88, 6e7, 0.02])
p_init1=np.array([-62.9, 6e7, 0.02])

#====================================================#
# The unexpected result using optimize.least_squares #
#====================================================#

result0 = least_squares(calc_residual, p_init0)
popt0 = result0.x
model_init0 = lorentzian(deltas, *p_init0)
model_opt0 = lorentzian(deltas, *popt0)

result1 = least_squares(calc_residual, p_init1)
popt1 = result1.x
model_init1 = lorentzian(deltas, *p_init1)
model_opt1 = lorentzian(deltas, *popt1)

fig, (ax0, ax1) = plt.subplots(1, 2)
fig.suptitle('Fitting results using optimize.least_squares')
fig.set_size_inches(8, 4)
ax0.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax0.plot(deltas, model_opt0, linewidth=0.5, marker='x', label='Fitted Model')
ax0.plot(deltas, model_init0, linewidth=1, label='Initial Model')
ax1.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax1.plot(deltas, model_opt1, linewidth=0.5, marker='x', label='Fitted Model')
ax1.plot(deltas, model_init1, linewidth=1, label='Initial Model')
ax0.legend()
fig.savefig('using_least_squares.pdf')

#====================================================#
#     The expected result using optimize.leastsq     #
#====================================================#

result0 = leastsq(calc_residual, p_init0)
popt0 = result0[0]
model_init0 = lorentzian(deltas, *p_init0)
model_opt0 = lorentzian(deltas, *popt0)

result1 = leastsq(calc_residual, p_init1)
popt1 = result1[0]
model_init1 = lorentzian(deltas, *p_init1)
model_opt1 = lorentzian(deltas, *popt1)

fig, (ax0, ax1) = plt.subplots(1, 2)
fig.suptitle('Fitting results using optimize.leastsq')
fig.set_size_inches(8, 4)
ax0.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax0.plot(deltas, model_opt0, linewidth=0.5, marker='x', label='Fitted Model')
ax0.plot(deltas, model_init0, linewidth=1, label='Initial Model')
ax1.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax1.plot(deltas, model_opt1, linewidth=0.5, marker='x', label='Fitted Model')
ax1.plot(deltas, model_init1, linewidth=1, label='Initial Model')
ax0.legend()
fig.savefig('using_leastsq.pdf')

#====================================================#
#    The expected result using optimize.curve_fit    #
#====================================================#

result0 = curve_fit(lorentzian, deltas, signal, p0=p_init0)
popt0 = result0[0]
model_init0 = lorentzian(deltas, *p_init0)
model_opt0 = lorentzian(deltas, *popt0)

result1 = curve_fit(lorentzian, deltas, signal, p0=p_init1)
popt1 = result1[0]
model_init1 = lorentzian(deltas, *p_init1)
model_opt1 = lorentzian(deltas, *popt1)

fig, (ax0, ax1) = plt.subplots(1, 2)
fig.suptitle('Fitting results using optimize.curve_fit')
fig.set_size_inches(8, 4)
ax0.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax0.plot(deltas, model_opt0, linewidth=0.5, marker='x', label='Fitted Model')
ax0.plot(deltas, model_init0, linewidth=1, label='Initial Model')
ax1.plot(deltas, signal, linewidth=0.5, marker='o', label='Data')
ax1.plot(deltas, model_opt1, linewidth=0.5, marker='x', label='Fitted Model')
ax1.plot(deltas, model_init1, linewidth=1, label='Initial Model')
ax0.legend()
fig.savefig('using_curve_fit.pdf')
```

### Error message

```shell
No error message.
```

### SciPy/NumPy/Python version and system information

```shell
1.11.3 1.22.3 sys.version_info(major=3, minor=9, micro=18, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /home/leeping/opt/miniconda3/envs/scipytest111/include
    lib directory: /home/leeping/opt/miniconda3/envs/scipytest111/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=0 PRESCOTT MAX_THREADS=128
    pc file directory: /home/leeping/opt/miniconda3/envs/scipytest111/lib/pkgconfig
    version: 0.3.21
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /home/leeping/opt/miniconda3/envs/scipytest111/include
    lib directory: /home/leeping/opt/miniconda3/envs/scipytest111/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK=0 NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP=0 PRESCOTT MAX_THREADS=128
    pc file directory: /home/leeping/opt/miniconda3/envs/scipytest111/lib/pkgconfig
    version: 0.3.21
  pybind11:
    detection method: pkgconfig
    include directory: /home/leeping/opt/miniconda3/envs/scipytest111/include
    name: pybind11
    version: 2.10.4
Compilers:
  c:
    commands: /croot/scipy_1696543286448/_build_env/bin/x86_64-conda-linux-gnu-cc
    linker: ld.bfd
    name: gcc
    version: 11.2.0
  c++:
    commands: /croot/scipy_1696543286448/_build_env/bin/x86_64-conda-linux-gnu-c++
    linker: ld.bfd
    name: gcc
    version: 11.2.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 0.29.36
  fortran:
    commands: /croot/scipy_1696543286448/_build_env/bin/x86_64-conda-linux-gnu-gfortran
    linker: ld.bfd
    name: gcc
    version: 11.2.0
  pythran:
    include directory: ../../_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.9/site-packages/pythran
    version: 0.13.1
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /home/leeping/opt/miniconda3/envs/scipytest111/bin/python
  version: '3.9'
```

**Comment by leeping** [comment]
Is anyone able to reproduce this issue?  Several of my students were also seeing it on their machines, so I don't think it's particular to my local machine.

**Comment by andyfaff** [comment]
I'm not sure what we could do in this situation. Do you see differences across different scipy versions for the same code? Otherwise it's not clear that there is a bug.

As you say there are a whole heap of things that could affect this, like starting parameters, or parameter scaling (the second parameter is a factor of 1e6 larger than the others). It's going to be difficult for devs to spend time looking into this specific problem unless there's clear evidence of a bug.

**Comment by leeping** [comment]
Thanks for the response.  I saw the same behavior in multiple SciPy versions ranging from v1.7 to v1.11.

I did some more testing and found the following.  

When using `least_squares(method='lm')` without providing variable scales via `x_scale`, an array of ones is passed as the last argument (`diag`) to `_minpack._lmdif`, which is the entry point of the MINPACK Levenberg-Marquardt algorithm.  (This is because `least_squares.py` computes `diag = 1 / x_scale` if `x_scale != 'jac'`).  During the second step of the optimization, the algorithm repeatedly cuts the step size until `xtol` is met, which leads to the poor quality-of-fit I described.  If `xtol` is reduced to 1e-10 then the correct fit is recovered for all algorithms.

When using `leastsq` or `curve_fit`, the default last argument to `_minpack._lmdif` is `diag=None`.  This results in significantly improved behavior of the optimization without a large number of step size cuts.  This is presumably because `_minpack._lmdif` is applying a default scaling to the parameters, (for example using the column norms of the Jacobian).  If I provide `diag=[1, 1, 1]` to calling `leastsq`, the poor behavior of `least_squares` is reproduced.

Therefore, I think the cause of the poor behavior in `least_squares` for my example was due to the default value of 1.0 for parameter scales.  However, `leastsq` and `curve_fit` applied a default parameter scaling that prevented the problem from appearing.  I think the latter behavior is more desirable and more robust.  Given that `leastsq` is already deprecated and `least_squares` is intended as a drop-in replacement, it would become an issue if a user of `leastsq` changes over to `least_squares` and gets very different results.  

In conclusion, I agree this isn't a bug per se, but it is arguably a performance regression, and I'd like to propose an enhancement where `least_squares(method='lm')` will pass `diag=None` to `_minpack._lmdif` as the default rather than an array of ones.  This is already the behavior when using `method='lm', x_scale='jac'` as the code explicitly sets `diag=None` in that case, so the enhancement could be implemented by setting the default value of `x_scale` to `jac` in the case of `method=='lm'`.  It is possible `x_scale='jac'` is also more robust for the other algorithms but I haven't tested this extensively.  

If there is agreement from the developers then we can change this issue from a bug report to an enhancement proposal.

**Comment by nish0699** [comment]
Hi @leeping,

We are a team of students from CentraleSupélec, we would like to work on this issue for our project for a class. We have reproduced the bug and will implement the enhancement suggested by you to compare the results. Also, analyse for any other fix possible. 
Can we work on this?

**Comment by melissawm** [comment]
@nish0699 we don't assign issues, so you are free to submit a PR with a fix if you want to. Please check [our contributing guide](https://docs.scipy.org/doc/scipy/dev/index.html) and feel free to reach out if you have questions. Cheers!

**Comment by leeping** [comment]
I'm not a SciPy maintainer but I would love to see your class implement this enhancement! :)
